---
title: "Examen - STAT3-2018"
author: "Denis Puthier"
date: "11/8/2018"
output:
  html_document:
    fig_caption: yes
    highlight: zenburn
    self_contained: yes
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: yes
  slidy_presentation:
    fig_caption: yes
    fig_height: 4
    fig_width: 6
    highlight: tango
    incremental: no
    keep_md: no
    self_contained: yes
    slide_level: 2
    smaller: yes
    theme: cerulean
    toc: yes
    widescreen: yes
  ioslides_presentation:
    colortheme: dolphin
    fig_caption: yes
    fig_height: 4
    fig_width: 6
    fonttheme: structurebold
    highlight: tango
    incremental: no
    keep_md: no
    slide_level: 2
    smaller: yes
    theme: cerulean
    toc: yes
    widescreen: yes
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 3
  beamer_presentation:
    colortheme: dolphin
    fig_caption: yes
    fig_height: 4
    fig_width: 6
    fonttheme: structurebold
    highlight: tango
    incremental: no
    keep_tex: no
    slide_level: 2
    theme: Montpellier
    toc: yes
font-import: http://fonts.googleapis.com/css?family=Risque
font-family: Garamond
transition: linear
bibliography: "document.bib"
csl: biomed-central.csl
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, eval = TRUE, cache = TRUE, message = FALSE, warning = FALSE,
                      comment = "")

## Options to display pretty numbers
library(knitr)
knit_hooks$set(inline = function(x) {
  prettyNum(x, big.mark=" ")
})
options(scipen = 6, digits = 3)

```


# About myelodysplastic syndromes and chronic myelomonocytic leukemia

In adults, circulating blood cells are formed in the bone marrow, through a process known as haematopoiesis. Blood cells arise from progenitors that pass through various differentiation steps before they become mature differentiated cells. They are various types of cancers of the blood-forming cells of the bone marrow. Chronic myelomonocytic leukaemia (CMML) is a type of leukaemia that arise from these progenitors and in which there are increased numbers of monocytes and immature blood cells (blasts) in the peripheral blood and bone marrow together with a set of abnormal looking cells (dysplasia) in at least one type of blood cell. Chronic myelomonocytic leukemia (CMML) are characterized by various mutations some of them targeting epigenetic modifiers and giving rise to aberrant DNA methylation. DNA methyltransferase inhibitors (DMTis) are used as a treatment for those disorders. Unfortunatly, patient response is highly variable (source [-@pmid25822018]).
In their article entitle "Specific molecular signatures predict decitabine response in chronic myelomonocytic leukemia", Meldi *et al* compared the transcriptome of bone marrow mononuclear cells (BM MNC) obtained from responders and non-responder patients to decitabine [-@pmid25822018]. They demonstrate that specific molecular signatures underlie response to decitabine in CMML. 

# Objective

This works aims at re-analysing data from Meldi *et al* using classical statistic tools.  

# Dataset

The dataset was obtained from recount2 database [-@pmid29043067; -@pmid28398307]. The recount2 database is an online resource consisting of RNA-seq gene and exon counts as well as coverage bigWig files for 2041 different studies. The dataset provided to students contains:

- An expression matrix (genes as rows and samples as columns) available [here](https://github.com/dputhier/SAOD---Statistical-analysis-of-omics-data/raw/gh-pages/examens/2018_stat3/exp_matrix.txt.gz).
- A matrix containing samples as rows and phenotype information as column. Avalaible [here](https://raw.githubusercontent.com/dputhier/SAOD---Statistical-analysis-of-omics-data/gh-pages/examens/2018_stat3/pheno_data.txt).
- A two columns matrix containing genesets. To be used for testing drug effects (see Question) section. Available [here](https://raw.githubusercontent.com/dputhier/SAOD---Statistical-analysis-of-omics-data/gh-pages/examens/2018_stat3/drug_gene_sets.txt)

All informations regarding this dataset can be obtained from SRA. The Bioproject ID of this study is (PRJNA260377)[https://www.ncbi.nlm.nih.gov/bioproject/].The Gene Expression Omnibus identifier is [GSE61162](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE61162).


# Questions

- **Descriptive statistics :** Using a set of well choosen diagrams and appropriate transformation (*e.g* addition of a pseudo-count and log2 transformation ) describe the basic features of the dataset. Try do highlight some potential biases. This step may be performed or not using ggplot2 library.

- **Unsupervized clustering with hierarchical clustering:** Using the whole dataset as input, compute a dendrogram displaying the hierarchical classification of the samples before feature selection (*i.e* gene selection). Discuss the results.

- **Detection differentially expressed genes:** run a statistical analysis of the read counts, in order to detect differentially expressed genes (**DEG**) between the two selected classes. Discuss the results. You may add a volcano plot to highlight the most differentially expressed genes.

- **Unsupervized clustering with hierarchical clustering after feature selection:** Using the selected genes as input, compute a dendrogram displaying the hierarchical classification of the samples after feature selection (*i.e* gene selection). Discuss the results.


- **Functional annotation:** Use the *gProfiler* R package to study the functional associations of the upregulated and downregulated genes, and provide a biological interpretation of these results.

- **Testing drug effects :** The 'drug_gene_sets.txt' contains a set of 38 gene sets (one per line). These gene sets correspond to the target genes that some bioactive molecules (*e.g* N-Benzylnaltrindole hydrochloride, BRD-K66792149, vorinostat...) may negatively regulate. Using a Fisher test whether an association exists between one or more of these lists and the list of genes upregulated in non-responsive patients. Conclude.

## Usage of the good practices

The report should contain the analysis of the provided data (see the *Questions* section). The expected format is a pdf or html file produced from a Rmarkdown file. During all your analysis you should take care to respect the following policies:

- **Tractability:** you and other people should be able to track the origin of all your results. For this, you need to keep a trace of each step of each analysis. 
    
- **Reproducibility:** other people should be able not only to trace the origin of your results, but also to reproduce them by themselves. NGS data analysis lends itself particularly well to reproducibility, since everything is done on computers and managed via software (tools and scripts). 
    
-  **Portability:** the analysis done on your computer should be reproducible on other computers as well. For this you need to ensure for instance that all paths are defined relative rather than absolute, and that the path definitions rely on platform-independent methods. 

-  **Spelling and grammar:** The report can be written in French or English. Sentences should be carefully written regarding the scientific meaning (no 'blah, blah, blah' just facts) and spelling and grammar. Manuscript not reaching the standard will be penalized.
        

# Format of the report

1. **Source document in Rmd**. The primary report is an R markdown document (.Rmd extension) which **must** contain all the code used to run the anlayses, as well as the main tables and figures produced by the analysis, and a text structured according to the common practice for scientific articles. 

    - The R code should be compliant with one of the following guidelines <https://style.tidyverse.org/> or <https://google.github.io/styleguide/Rguide.xml>.
    - The R code should be properly documented. 

2. **Compiled report (html or pdf)**. This report should look like a small scientific article (see structure hereafter) with Figures, Tables, and interpretation of the results.  The **R** code should not be displayed in the compiled report (set the knitr option `r{echo=FALSE}` when generating the last version). Think of your report as a document written for a biologists who wants to understand the approach and the results, but is not interested by the technical details of the R programming. 

# Structure of the report

In total, the report should not exceed 7-8 pages, including figures, but without counting the bibliographic references and appendices (for which there is no limit).

1. **Introduction:** a brief summary (5-10 lines) of the biological context (the disease, the transcriptome), the biological question addressed in the report, and the general approach envisaged to answer these questions. 

2. **Material and Methods:** a summary of the bioinformatics / statistical methods and libraries used for the analysis (1/2 to 1 page), with a brief explanation about which tool was used to do what, and links to the official web page or publication about the tool. If required, additional details about the methods and parameters can be provided in appendix. 

3. **Data description:** a brief description of the data source (with link to the GEO record), its content (how many samples, how many groups, ...) and the data type (paired-ends or single-end, ...).

4. **Results and discussion:** the results should be presented and discussed together in this section. This section contains the main figures and tables that are used to interpret the results. Additional figures and tables can be provided as appendices, or as separate files (for example full result tables with all the genes, ...).

5. **Conclusion and perspective** (~1/2 page): summarize the results, show in what they did -- or did not -- enable you to answer the initial questions of the introduction, and add some perspectives about possible future extensions of the work presented here. 




**Format:** the report can be submitted in either html or pdf format. The original Rmd file that was used to generate the report must be submitted together with the report. This file should allow the teachers to reprodue the analysis on their computers. 


# Evaluation criteria

The evaluation of your report will be based on multiple criteria.

1. Can we reproduce your results using your Rmd file?
2. Did you clearly formulate the biological question, and the relationship between these questions and the bioinformatics approaches used to answer them?
3. Relevance of the results and your interpretation. This last point is of particular importance. 
4. Clarity of the text.
5. Clarity of the code. 


****************************************************************
# References



